### ISSUE TITLE
Allow entities to be active outside the set 'mapgen limit'

### ISSUE THREAD

**Issue by paramat** [minetest/minetest#6984]
A server owner is experiencing a server attack where massive amounts of mapblocks are somehow being generated outside a certain limit, the limit is enforced by a mod that automatically teleports players. Currently the only defense seems to be using the 'mapgen limit' setting to enforce a reduced limit.
The problem is that entities are now being removed outside the 'mapgen limit', even though existing world outside the reduced limit is still visible and accessible to players.

The intention of 'mapgen limit', which was hmmmm's intention also, is that it only prevents generation of new terrain outside a set limit, it should not do anything else. The limit is often changed and reduced for various reasons and the world outside is deliberately visible and accessible to players. It should function normally in all ways.

Unfortunately it seems that somewhere in the code the 'mapgen limit' is being treated as 'the edge of the world' and some things are being prevented outside it, this is the wrong approach.
See 'objectpos_over_limit()':
```
inline bool objectpos_over_limit(v3f p)
{
	const float max_limit_bs = MAX_MAP_GENERATION_LIMIT * BS;
	return p.X < -max_limit_bs ||
		p.X >  max_limit_bs ||
		p.Y < -max_limit_bs ||
		p.Y >  max_limit_bs ||
		p.Z < -max_limit_bs ||
		p.Z >  max_limit_bs;
}
```
Objects are always allowed up to 'MAX_MAP_GENERATION_LIMIT' (31000).
Somewhere entities are being removed outside the limit.
@Ezhh

**Comment by paramat** [comment]
Quick fix is https://github.com/minetest/minetest/blob/master/src/mapgen/mapgen.cpp#L1101
```
	// Effective mapgen limit, in blocks
	// Uses same calculation as ServerMap::blockpos_over_mapgen_limit(v3s16 p)
	s16 mapgen_limit_b = rangelim(mapgen_limit,
		0, MAX_MAP_GENERATION_LIMIT) / MAP_BLOCKSIZE;
```
Change to:
`s16 mapgen_limit_b = MAX_MAP_GENERATION_LIMIT / MAP_BLOCKSIZE;`

**Comment by Ezhh** [comment]
Removal of entities in mapblocks that were never generated is of course good, but I believe they should remain present in mapblocks that have previously been generated, regardless of the current mapgen_limit. 

I will be very appreciative if it can be changed to work this way.

**Comment by paramat** [comment]
Related code https://github.com/minetest/minetest/blob/fad263dec9e1fa7ae0886f768e22d8ee74e8553b/src/serverenvironment.cpp#L571
https://github.com/minetest/minetest/blob/de2c40c8fc189eeeeba010a0aaa1d28aed588c68/src/content_sao.cpp#L403

**Comment by Wuzzy2** [comment]
This is not a bug, it's a feature. There is a reason we decided to remove entities out of map bounds. Because it's easy for i.e. falling nodes and item entities to fall outside the world and never get cleared. This is, of course, a waste of resources. Currently, item entities and falling nodes are removed outside the map bounds.

> The intention of 'mapgen limit', which was hmmmm's intention also, is that it only prevents map generation outside a set limit, it should not do anything else. The limit is often changed and reduced for various reasons and the world outside is deliberately visible and accessible to players. It should function normally in all ways.

Wait, what? So if you reduce the limit, the map does not actually shrink if it has already been generated?
Damn. This makes things more complicated.

Sorry, paramat, but I think a simple “revert this commit” won't work here. If you do this, items and falling nodes and other entities will be able to fall out of the world again without getting cleared. And you can re-open #5824 then.

**Comment by paramat** [comment]
Yes there may be issues caused by this. It seems better to stop entities from being able to pass into "ignore" nodes (ungenerated mapblocks), just like players. Then the entities are limited by the edge of the generated world as they should be, instead of by a set cuboid limit and a per-step co-ordinate checking function as now which is a messy solution.
Does anyone know where in the code entities are allowed to enter ignore? and why? I'll look but just asking.

Usually the mapgen limit is set on world creation and never changed, in this common case there never is world outside the limit.
When server owners choose to reduce the limit it seems they usually want the generated world outside to remain functional.

**Comment by paramat** [comment]
However, entities falling into ignore was not a big problem and is worth it if it allows servers to defend themselves against these attacks. So i will write a PR for backporting to 0.4.17 and we can then tackle entities another way, hopefully by not letting them enter ignore nodes.

**Comment by benrob0329** [comment]
Ignore is walkable, shouldn't entities stop on ignore if collisions check for walkable?

**Comment by Wuzzy2** [comment]
Good point, benrob0329! It's strange that players can collide with walkable blocks and `ignore`, but objects are like ghosts.
I know, block collision can be disabled on purpose for entities. But how about forcing objects to always collide with `ignore` no matter what, as a special case?

Also, if an object ended up in `ignore` somehow (maybe by teleporting, or a borked map), maybe delete the entity (after 1 second or so, to give it a “final chance” to recover).

**Comment by paramat** [comment]
PR #6998

**Comment by Fixer-007** [comment]
There needs to be option in config to disable/enable this behaviour.

**Comment by paramat** [comment]
Many entities are 'physical = false' so won't collide with walkable nodes, so a specific check for ignore is needed.

**Comment by paramat** [comment]
The 'builtin:item' entity can be dropped into ignore, it falls through but not smoothly, the code seems to need a check for ignore here:
https://github.com/minetest/minetest/blob/d693f95fc3c9b2d487283fc16dc8988c61a94803/builtin/game/item_entity.lua#L151

**Comment by paramat** [comment]
PR #6999 stops item entites moving in ignore, which will help the situation if #6998 is merged.

Mods and Lua code that create and move entities have a responsibility to prevent these entities from moving in ignore, this is not the responsibility of the engine. Ignore is not walkable so collision checks in engine cpp code are not relevant.

**Comment by paramat** [comment]
#6999 merged.

**Comment by paramat** [comment]
If #7024 is merged this PR can then be merged.
I will make sure boats cannot enter ignore.
@SmallJoker i assume carts will not enter ignore becasue there are no rails?

**Comment by paramat** [comment]
2 more PRs as part of this: #7024 #7032

**Comment by paramat** [comment]
#7024 #7032 merged.
Just #6998 remains.

**Comment by paramat** [comment]
https://github.com/minetest/minetest_game/pull/2044 merged.
